src/main/scala/scalismo/faces/segmentation/LoopyBPSegmentation.scala

```diff
@@ -0,0 +1,478 @@
+ *  limitations under the License.
+ */
+
+
+package scalismo.faces.segmentation
+
+import java.io.File
+
+import scalismo.faces.color.{ColorDistribution, GaussDist, RGB, UniformDist}
+import scalismo.faces.image.{ImageBuffer, PixelImage}
+import scalismo.faces.gui.GUIBlock._
+import scalismo.faces.gui.{GUIFrame, ImagePanel}
+import scalismo.faces.io.PixelImageIO
+import scalismo.utils.Random
+
+import scala.collection.immutable.IndexedSeq
+
+
+// an implementation of loopy belief propagation for segmentation
+object LoopyBPSegmentation {
+
+
+  // the directions for message passing
+  private sealed trait Direction {
+    def toInt: Int
+  }
+
+  private object Direction {
+    val allDirections = IndexedSeq(Right, Left, Down, Up)
+
+    def fromIndex(index: Int): Direction = index match {
+      case 0 => Right
+      case 1 => Left
+      case 2 => Down
+      case 3 => Up
+      case _ => throw new RuntimeException("unknown direction index")
+    }
+  }
+
+  private case object Right extends Direction {
+    override def toInt: Int = 0
+  }
+
+  private case object Left extends Direction {
+    override def toInt: Int = 1
+  }
+
+  private case object Down extends Direction {
+    override def toInt: Int = 2
+  }
+
+  private case object Up extends Direction {
+    override def toInt: Int = 3
+  }
+
+  /** the segmentation label */
+  case class Label(value: Int) extends AnyVal
+
+  /** binary distribution, local information */
+  type BinaryLabelDistribution = PixelImage[(Label, Label) => Double]
+
+  private type MessageFieldBuf = ImageBuffer[Array[LabelDist]]
+
+  /** MRF segmentation with Loopy Belief Propagation
+    * Builds simple Gaussian color models for each label iteratively
+    *
+    * @param image image to segment
+    * @param labelInit initial label where known
+    * @param numLabels number of labels in use */
+
+  def segmentImage(image: PixelImage[RGB],
+                   labelInit: PixelImage[Option[Label]],
+                   binaryDistribution: PixelImage[(Label, Label) => Double],
+                   numLabels: Int, numIterations: Int,
+                   gui: Boolean = false): PixelImage[LabelDist] = {
+    require(labelInit.values.forall(i => i.forall(_.value < numLabels)))
+
+    // init: color models and start labels
+    val initialSegmentation = labelInit.map{init => LabelDist.fromSingleLabel(init.getOrElse(Label(numLabels-1)), numLabels, 0.01)}
+    var colorDists = estimateColorDists(image, initialSegmentation)
+
+    // initialize message field: constant messages, unknown distributions yet
+    val messageField = ImageBuffer.makeInitializedBuffer(image.width, image.height)(Array.fill(Direction.allDirections.length)(LabelDist.constant(numLabels)))
+
+    // calculate local messages: according to color distributions
+    var localMessages = image.map{c => LabelDist(colorDists.map{dist => dist.evaluate(c)})}.map{_.normalized}
+    val counterLabel = label("init")
+    val guiFrame = if (gui) {
```

Comment:
I would argue to either remove this debugging / visualization code from the main function. If you really want to keep it then I would argue to place it in a VisualSegmentation class or so.

Reply:
that would lead to code duplication - since it is visualizing core components of the segmentation.
I would like to keep it since I want to use it in the tutorial

Reply:
I can not say that i have already a fully working solution in mind. I thought about creating a visualization class that uses this class for the calculation. Perfoming one iteration at the time and visualizing the result afterwards. Maybe we can open an issue for that to decouple it later.

Reply:
it is strongly coupled since it does not only visualize the final result of the calculation but also the messages and the beliefs...